Add summary display style to Reactions block

pfefferle opened the pull request:
Fixes #1815

## Proposed changes:

Add a new "Compact" display style to the Reactions block that shows reaction counts inline without avatar facepiles. Both styles use the same underlying Reactions component - the Compact style simply hides the avatars in the facepile row and modal dropdown.

### Features:
- **Facepile style** (default): Shows avatar facepiles with reaction counts
- **Compact style**: Shows only reaction count buttons inline, without avatars
- Automatically defaults to Compact style when WordPress "Show Avatars" setting is disabled (Settings → Discussion)
- When avatars are disabled, the Facepile style option is hidden from the block styles panel

| Compact | Facepile |
|---|---|
| <img width="1553" height="1263" alt="Compact style screenshot" src="https://github.com/user-attachments/assets/d669d4d2-3134-4a15-b53a-a3b5eeecadf8" /> | <img width="1553" height="1263" alt="Facepile style screenshot" src="https://github.com/user-attachments/assets/542a0f15-cbde-4703-8b2e-d3abe3333b9c" /> |

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:

1. Go to the block editor and insert a "Fediverse Reactions" block
2. In the block toolbar, click the styles icon to switch between "Facepile" and "Compact" styles
3. When "Compact" style is selected, verify:
   - No avatar facepiles are shown
   - Reaction buttons are displayed inline
   - Clicking a button opens a dropdown without avatars
4. Test that the block renders correctly on the frontend with both styles
5. Go to Settings → Discussion and disable "Show Avatars"
6. Insert a new Reactions block and verify it defaults to "Compact" style
7. Verify the "Facepile" style option is hidden when avatars are disabled

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [x] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Add compact display style to Reactions block that hides avatars.

</details>

pfefferle committed:
Add summary display style to Reactions block

Introduce a new "Summary" display style for the Reactions block that
shows reaction counts as text instead of avatars. This is useful for
users who prefer a more compact display or have avatars disabled.

Key changes:
- Add block styles (Facepile/Summary) via block.json
- Add configurable settings for summary style: separator character,
  show/hide comments count, and show/hide empty reactions
- Sync displayStyle attribute with WordPress block style selection
- Default to summary style when WordPress show_avatars option is disabled
- Refactor render.php to share common wrapper code between styles

matticbot committed:
Add changelog

pfefferle committed:
Hide empty headings in Reactions block

obenland commented on src/reactions/edit.js:
```diff
@@ -36,20 +39,102 @@ const DUMMY_REACTIONS = {
+		const hasStyleClass = className?.includes( 'is-style-' );
+		if ( ! showAvatars && ! hasStyleClass ) {
+			setAttributes( {
+				className: ( className ? className + ' ' : '' ) + 'is-style-summary',
```
We add `clsx` as a dependency in `add/router`, maybe we could move that up into this PR and use it here.

obenland commented on src/reactions/edit.js:
```diff
@@ -36,20 +39,102 @@ const DUMMY_REACTIONS = {
+	const classNameStyle = className?.includes( 'is-style-summary' ) ? 'summary' : 'facepile';
+	if ( classNameStyle !== displayStyle ) {
+		setAttributes( { displayStyle: classNameStyle } );
+	}
```
This looks like good feedback

obenland commented on src/reactions/edit.js:
```diff
@@ -65,13 +150,45 @@ export default function Edit( { __unstableLayoutClassNames } ) {
+							checked={ showComments }
+							onChange={ ( value ) => setAttributes( { showComments: value } ) }
+							__nextHasNoMarginBottom
+						/>
```
I'd expect the dummy block in the editor to update when this control gets toggled, and remove the comments.

obenland commented on src/reactions/edit.js:
```diff
@@ -65,13 +150,45 @@ export default function Edit( { __unstableLayoutClassNames } ) {
+							checked={ showEmpty }
+							onChange={ ( value ) => setAttributes( { showEmpty: value } ) }
+							__nextHasNoMarginBottom
+						/>
```
Same as above, let's remove reactions that don't have a count, in the dummy that would be quotes.

Should we make this available to facepile style as well?

obenland commented on src/reactions/edit.js:
```diff
@@ -65,13 +150,45 @@ export default function Edit( { __unstableLayoutClassNames } ) {
+							onChange={ ( value ) => setAttributes( { separator: value } ) }
+							__next40pxDefaultSize
+							__nextHasNoMarginBottom
+						/>
```
I'm not in love with this one. Is it necessary to have that?

It feels like it could be simplified with a 20px margin between reaction types, no separator.

obenland commented on src/reactions/render.php:
```diff
@@ -38,133 +55,167 @@
+
+	// Don't render if there are no items to show.
+	if ( empty( $counts ) ) {
+		ob_end_clean();
```
```suggestion
		ob_end_clean();
		echo '<!-- Reactions block: No reactions found. -->';
```

obenland commented on src/reactions/render.php:
```diff
@@ -206,14 +257,56 @@ class="reaction-label wp-element-button"
+						data-wp-bind--src="context.item.avatar"
+						data-wp-on--error="callbacks.setDefaultAvatar"
+						src=""
+					/>
```
We should not show this when show_avatars is false

obenland commented on src/reactions/render.php:
```diff
@@ -206,14 +257,56 @@ class="reaction-label wp-element-button"
 		</div>
```
If feel like we shouldn't show avatars in the facepile if show_avatars is false.

pfefferle replied:
how would that look like 🤔

the whole idea of this PR was to not use facepiles when avatars are disabled!?

obenland replied:
If we don't just want to show the labels, which I'd understand, we could not make the default style available.

Thinking further on this, should we still offer the modals with the user lists in the summary style?

obenland replied:
maybe the difference between the two styles could just be CSS and whether the avatars are displayed or not.

pfefferle replied:
it was a request to have that style even with avatars enabled.

pfefferle replied:
btw. I changed the code to auto add the summary version to the post by default when avatars are disabled.

pfefferle replied:
This is only an issue if you add it on purpose.

obenland commented on src/reactions/edit.js:
```diff
@@ -65,13 +150,45 @@ export default function Edit( { __unstableLayoutClassNames } ) {
 			/>
-			<Reactions postId={ getCurrentPostId() } fallbackReactions={ DUMMY_REACTIONS } />
+			{ isSummaryStyle ? (
+				<Summary separator={ separator } />
```
Let's pass it the post ID, like in `Reactions` and display the actual reactions if available.

obenland replied:
It looks like this is still missing.

obenland commented on src/reactions/render.php:
```diff
@@ -38,133 +55,167 @@
-);
+		$counts[ $_type ] = array(
+			'count' => $count,
+			'label' => $type_object['label'],
```
This seems to be missing singular labels.

```php
		$label = sprintf(
			_n(
				$type_object['count_single'],
				$type_object['count_plural'],
				$count,
				'activitypub'
			),
			number_format_i18n( $count )
		);
```

pfefferle committed:
Address PR feedback for reactions summary style

- Use clsx for className handling
- Move setAttributes sync to useEffect to prevent infinite re-renders
- Make Summary component respect showComments and showEmpty attributes
- Add singular/plural labels in render.php using _n()
- Use ob_end_clean() with HTML comment on empty reactions
- Remove separator option - use CSS margin instead
- Don't show avatars in facepile/modal when show_avatars is false

pfefferle committed:
Merge branch 'trunk' into add/reactions-summary-style

pfefferle committed:
Update src/reactions/render.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Remove duplicate comment in reactions render

Eliminated a redundant HTML comment output when no reactions are found in the summary block.

pfefferle committed:
Merge branch 'trunk' into add/reactions-summary-style

obenland commented on src/reactions/render.php:
```diff
@@ -38,140 +54,179 @@
+		if ( $attributes['showEmpty'] || $comment_count > 0 ) {
+			$counts['comments'] = array(
+				'count' => $comment_count,
+				'label' => _n( 'Comment', 'Comments', $comment_count, 'activitypub' ),
```
The way label is being rendered below, we need to include the comment count into the label string.

<img width="301" height="104" alt="Image" src="https://github.com/user-attachments/assets/8bc5d40b-5baa-4014-a2c4-b1c27c17aca9" />

obenland commented on src/reactions/render.php:
```diff
@@ -38,140 +54,179 @@
+		$label = sprintf(
+			_n(
+				$type_object['count_single'],
+				$type_object['count_plural'],
```
Hm, that's unfortunate. Looks like the translation strings only accept numeric values, so we can't pass a string here. As a result, the number will not be bold here, as it is in the editor. I also just noticed there's a difference in capitalization.

<img width="301" height="104" alt="Image" src="https://github.com/user-attachments/assets/cf9e5b6c-4a1c-42e5-acab-e7d1fe57f81e" />

<img width="348" height="131" alt="Image" src="https://github.com/user-attachments/assets/382d4a93-2a65-4924-a729-a8e404bb5d08" />

pfefferle committed:
Fetch real reaction counts in Summary component

Update Summary component to fetch actual reaction counts from the API,
similar to how the Reactions component works. Falls back to dummy data
if no real reactions exist or in the Site Editor.

pfefferle committed:
Unify facepile and summary styles into single Reactions component

Both styles now use the same underlying code path. The summary style
hides avatars in both the facepile row and modal dropdown by checking
the displayStyle attribute. This simplifies the codebase by removing
the separate Summary component.

pfefferle committed:
Show avatars only in facepile display style

Updated the ReactionList component to render user avatars only when the display style is set to 'facepile', instead of for all styles except 'summary'. This ensures avatars are not shown in other display modes, aligning the UI with intended design.

pfefferle committed:
Wrap reactions in container and adjust font size

The Reactions component now renders its content inside a div with class 'activitypub-reactions' instead of a fragment, improving structure and styling. Additionally, the font size for .reaction-name is set to 0.875rem for better visual consistency.

pfefferle committed:
Use CSS variable for reaction name font size

Updated the .reaction-name class to use the --wp--preset--font-size--small CSS variable with a fallback, improving theme compatibility. Regenerated CSS build files to reflect this change.

pfefferle committed:
Update reaction label font size and summary heading margin

Changed the reaction label font size to use the CSS variable --wp--preset--font-size--small for better consistency with theme settings. Added a bottom margin to headings in the summary style for improved spacing. Updated compiled CSS files accordingly.

pfefferle commented:
@obenland with all the feedback above and working on the fixes, I now think it really makes sense to re-use the Reaction Block for both!

> maybe the difference between the two styles could just be CSS and whether the avatars are displayed or not.

It is not only CSS, but the current version is a somehow Avatarless version of the Reactions Block. The `Summary` style simply disables the facepiles (avatars), but keeps the dropdown and shows the "buttons" inline.

| Summary | Facepiles |
|---|---|
| <img width="1553" height="1263" alt="Screenshot 2025-12-12 at 09 19 59" src="https://github.com/user-attachments/assets/d669d4d2-3134-4a15-b53a-a3b5eeecadf8" /> | <img width="1553" height="1263" alt="Screenshot 2025-12-12 at 09 20 03" src="https://github.com/user-attachments/assets/542a0f15-cbde-4703-8b2e-d3abe3333b9c" /> |

pfefferle committed:
Merge branch 'trunk' into add/reactions-summary-style

pfefferle committed:
Default reactions block style based on avatar setting

The reactions block now defaults to the 'facepile' style if avatars are enabled and 'summary' if avatars are disabled, unless a style is explicitly set. The block registration also removes the 'facepile' style when avatars are disabled. This ensures consistent display behavior in both editor and frontend for auto-inserted blocks.

pfefferle committed:
Rename 'summary' style to 'compact' for reactions block

Replaces the 'summary' display style with 'compact' throughout the reactions block codebase, including block registration, PHP rendering, JS logic, styles, and tests. The 'compact' style now hides avatars, and is used as the default when avatars are disabled. Updates all related documentation, selectors, and test cases to reflect the new naming.

pfefferle committed:
Add compact display style to Reactions block

Introduces a new compact display style for the Reactions block that hides user avatars, providing a more streamlined appearance.

obenland commented on src/reactions/render.php:
```diff
@@ -38,6 +51,21 @@
+if ( ! $has_style_class ) {
+	$default_style              = get_option( 'show_avatars', true ) ? 'facepile' : 'compact';
+	$attributes['className']    = trim( ( $attributes['className'] ?? '' ) . ' is-style-' . $default_style );
+	$attributes['displayStyle'] = $default_style;
```
```suggestion
	$attributes['className']    = trim( ( $attributes['className'] ?? '' ) . ' is-style-' . $default_display_style );
	$attributes['displayStyle'] = $default_display_style;
```

obenland commented on src/reactions/render.php:
```diff
@@ -38,6 +51,21 @@
+}
+
+$is_compact   = 'compact' === $attributes['displayStyle'];
+$show_avatars = ! $is_compact;
```
```suggestion
$show_avatars = 'facepile' === $attributes['displayStyle'];
```

obenland commented on includes/class-blocks.php:
```diff
@@ -84,7 +84,18 @@ public static function register_blocks() {
+					'isDefault' => true,
+				),
+			);
+		}
```
This doesn't seem to work for me. When I insert a new Reactions block with avatars disabled, it still default to facepile for me. Maybe something we need to do on the JS side

pfefferle replied:
This is mainly to deactivate the "Facepile Style" in the sidebar. I have not tested your case tbh. 🫣

obenland reviewed (commented):
@pfefferle This update is excellent.

The one question remaining for me is wether we should force compact display when show_avatar is false, regardless of choice of style, particularly for blocks that are already inserted.

pfefferle commented:
> The one question remaining for me is wether we should force compact display when show_avatar is false, regardless of choice of style, particularly for blocks that are already inserted.

I think either way, we should have a consistent handling over all blocks.

pfefferle committed:
Apply suggestion from @obenland

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Apply suggestion from @obenland

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Fix avatar display logic in reactions render

Corrects the logic for determining when to show avatars by checking if the display style is 'facepile' instead of relying on 'compact' style. This ensures avatars are shown or hidden according to the selected display style.

pfefferle committed:
Align JSDoc param formatting in reactions components

Standardized the alignment of JSDoc @param and @property tags across edit.js, reactions.js, and view.js for improved readability and consistency.

pfefferle committed:
Merge branch 'trunk' into add/reactions-summary-style

obenland commented on includes/class-blocks.php:
```diff
@@ -84,7 +84,18 @@ public static function register_blocks() {
+					'label'     => __( 'Compact', 'activitypub' ),
+					'isDefault' => true,
+				),
+			);
```
```suggestion
			$reactions_args['styles'] = array();
```

Removing all styles when we don't show avatars might be the best way forward?

With only Compact, the block still displays a Styles section but you can't change it, which feels awkward. With empty styles it defaults to compact without showing the styles section.

<img width="287" height="120" alt="Image" src="https://github.com/user-attachments/assets/a2899367-51f9-481d-9d87-5fa7d3f1831c" />

obenland reviewed (commented):
I think that's all I have. Not sure if I covered all scenarios in testing, but it should be pretty easy to fix/update if there's more we find later.

pfefferle committed:
Update includes/class-blocks.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Merge branch 'trunk' into add/reactions-summary-style

kaffeeringe commented:
Cool! Now it would be nice to have a comments count block in the same style. ;-)
